src/rust/src/x509.rs

```diff
@@ -539,6 +540,509 @@ fn load_der_x509_certificate(py: pyo3::Python<'_>, data: &[u8]) -> PyAsn1Result<
+    fn fingerprint<'p>(
+        &self,
+        py: pyo3::Python<'p>,
+        algorithm: pyo3::PyObject,
+    ) -> pyo3::PyResult<&'p pyo3::PyAny> {
+        let hashes_mod = py.import("cryptography.hazmat.primitives.hashes")?;
+        let h = hashes_mod.getattr("Hash")?.call1((algorithm,))?;
+        h.call_method1("update", (self.public_bytes_der().as_slice(),))?;
+        h.call_method0("finalize")
+    }
+
+    #[getter]
+    fn signature_algorithm_oid<'p>(&self, py: pyo3::Python<'p>) -> pyo3::PyResult<&'p pyo3::PyAny> {
+        let x509_module = py.import("cryptography.x509")?;
+        x509_module.call_method1(
+            "ObjectIdentifier",
+            (self.raw.borrow_value().signature_algorithm.oid.to_string(),),
+        )
+    }
+
+    #[getter]
+    fn signature_hash_algorithm<'p>(
+        &self,
+        py: pyo3::Python<'p>,
+    ) -> pyo3::PyResult<&'p pyo3::PyAny> {
+        let oid = self.signature_algorithm_oid(py)?;
+        let oid_module = py.import("cryptography.hazmat._oid")?;
+        let exceptions_module = py.import("cryptography.exceptions")?;
+        match oid_module.getattr("_SIG_OIDS_TO_HASH")?.get_item(oid) {
+            Ok(v) => Ok(v),
+            Err(_) => Err(pyo3::PyErr::from_instance(exceptions_module.call_method1(
+                "UnsupportedAlgorithm",
+                (format!(
+                    "Signature algorithm OID:{} not recognized",
+                    self.raw.borrow_value().signature_algorithm.oid
+                ),),
+            )?)),
+        }
+    }
+
+    #[getter]
+    fn signature(&self) -> &[u8] {
+        self.raw.borrow_value().signature_value.as_bytes()
+    }
+
+    #[getter]
+    fn tbs_certlist_bytes<'p>(&self, py: pyo3::Python<'p>) -> &'p pyo3::types::PyBytes {
+        let b = asn1::write_single(&self.raw.borrow_value().tbs_cert_list);
+        pyo3::types::PyBytes::new(py, &b)
+    }
+
+    fn public_bytes<'p>(
+        &self,
+        py: pyo3::Python<'p>,
+        encoding: &pyo3::PyAny,
+    ) -> pyo3::PyResult<&'p pyo3::types::PyBytes> {
+        let encoding_class = py
+            .import("cryptography.hazmat.primitives.serialization")?
+            .getattr("Encoding")?;
+
+        let result = asn1::write_single(self.raw.borrow_value());
+        if encoding == encoding_class.getattr("DER")? {
+            Ok(pyo3::types::PyBytes::new(py, &result))
+        } else if encoding == encoding_class.getattr("PEM")? {
+            let pem = pem::encode_config(
+                &pem::Pem {
+                    tag: "X509 CRL".to_string(),
+                    contents: result,
+                },
+                pem::EncodeConfig {
+                    line_ending: pem::LineEnding::LF,
+                },
+            )
+            .into_bytes();
+            Ok(pyo3::types::PyBytes::new(py, &pem))
+        } else {
+            Err(pyo3::exceptions::PyTypeError::new_err(
+                "encoding must be an item from the Encoding enum",
```

Comment:
The encoding enum has tons of stuff, we should probably just say it must be DER or PEM from encoding.

Reply:
(This should be a followup PR)